### PR TITLE
Checkout: Show empty cart view when cart is empty

### PR DESCRIPTION
**Pull request by sirbrillig**
#### Changes proposed in this Pull Request

This adds a new view that renders in checkout when the cart is empty. Since typically a customer will be redirected to the plans page when the cart is empty, this should normally never be seen. However, if there is an error displayed on the screen which prevents the redirect, then this view will be shown instead of the loading animation.

Before:

Screen Shot 2020-11-04 at 7 51 00 PM

After:

Screen Shot 2020-11-04 at 7 50 37 PM

Fixes https://github.com/Automattic/wp-calypso/issues/45182

#### Testing instructions

- Add a plan to your cart and visit checkout. Verify that you see the loading animation until the cart finishes loading, and that it then renders checkout with the plan in the cart. Verify that you never see the empty cart view.
- Remove the product from your cart. Verify that you see the pending state on the form's buttons until the change completes, and that you are redirected to the plans page.  Verify that you never see the empty cart view.
- Visit a URL that adds a non-existent product like `/checkout/example.com/abcd`. Verify that you see the loading animation and then you see the new empty cart view with the appropriate error message.

### CODE REVIEW

**Comment by sirbrillig** [comment]
Per the original issue, we should add an analytics call to this page and the button, but the UI and UX can be reviewed now.

**Comment by fditrapani** [comment]
Thanks for putting this together. It works well for this case:

> Visit a URL that adds a non-existent product like /checkout/example.com/abcd. Verify that you see the loading animation and then you see the new empty cart view with the appropriate error message.

But when you removing a product from the cart, can we should show the updating cart state until it redirects? 

image

File: client/my-sites/checkout/composite-checkout/composite-checkout.tsx [hunk 1]
```diff
@@ -630,6 +640,32 @@ export default function CompositeCheckout( {
 		} );
 	}
 
+	if ( responseCart.products.length < 1 && ( areThereErrors || ! isInitialCartLoading ) ) {
+		const goToPlans = () => ( siteSlug ? page( `/plans/${ siteSlug }` ) : page( '/plans' ) );
```
Comment (fditrapani) [inline client/my-sites/checkout/composite-checkout/composite-checkout.tsx:644]: Can we add an event to see how many people click this?

Comment (sirbrillig) [reply client/my-sites/checkout/composite-checkout/composite-checkout.tsx]: Done!

**Comment by sirbrillig** [comment]
> But when you removing a product from the cart, can we should show the updating cart state until it redirects?

Yeah, that's a good point. I think I fixed that now.

**Comment by fditrapani** [comment]
I think something broke with the update. I'm seeing the empty cart screen right after adding a plan to my cart, then it shows the loader, then the cart. I also still see the empty cart after removing a product.

**Comment by fditrapani** [comment]
I confirmed that the `calypso_checkout_composite_empty_cart_clicked` event is firing :)

**Comment by sirbrillig** [comment]
Oops! Yes, sorry about that! I definitely messed up the logic. Try it now.

**Comment by fditrapani** [comment]
Thanks for the update — almost there! The only thing is when I remove a product, I'm seeing the loading state vs. the updating cart state:

image

**Comment by sirbrillig** [comment]
Good point! Unfortunately, it turns out this is pretty tricky to arrange because of how we store the shopping cart. Here's the order of events:

1. Click the confirm button to remove the product from the cart.
2. The product is removed from the local cache of the shopping cart and the cache is marked as invalid.
3. The invalid cache sends a request to the shopping-cart endpoint to get an updated version of the cart to match our local cache.
4. Meanwhile, the shopping cart manager gives the list of products back to the UI (the list is empty) and tells it that the cart is updating.
5. Normally if the cart is updating, we display the "updating" state that you described above, but if the cart is empty, then we display the loading state instead because the UI isn't set up to show an empty cart (actually, that's what this PR is meant to help resolve).

I think what you're suggesting, and it's a good suggestion, is that when the cart is updating we should show the _previous_ version of the cart until the update completes. This is certainly possible and may improve other UX interactions as well, but IMO it's outside the scope of this PR since it doesn't have anything to do with the empty cart view, and the behavior you reported is also happening in production. I created https://github.com/Automattic/wp-calypso/issues/47163 to track that.

**Comment by fditrapani** [comment]
Cool, thanks for the breakdown of events and creating the extra ticket. In that case then, we're good to go!

**Review by fditrapani — APPROVED** [review]

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/5123313

Thank you @sirbrillig for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/5123313).

**Pull request merged**